File: src/content/3.5/code/scala/snippet12.scala
implicit def readerMonad[E] = new Monad[Reader[E, ?]] {
  def pure[A](x: A): Reader[E, A] =
    Reader(e => x)
  
  def flatMap[A, B](ra: Reader[E, A])(k: A => Reader[E, B]): Reader[E, B] =
    Reader(e =>
      runReader(k(runReader(ra)(e)))(e))
}

File: src/content/3.5/code/ocaml/snippet12.ml
module ReaderMonad (T : sig
  type t
end) : Monad_Bind with type 'a t = (T.t, 'a) reader = struct
  type 'a t = (T.t, 'a) reader

  let ( >>= ) ra k =
    Reader (fun e -> run_reader (k (run_reader ra e)) e)
  let return a = Reader (fun e -> a)
end
